glean_parser/util.py

```diff
@@ -399,6 +420,19 @@ def report_validation_errors(all_objects):
     return found_error
 
 
+def remove_output_params(d, output_params):
+    """
+    Removes output-only params, such as "defined_in",
+    in order to validate the output against the input schema.
+    """
+    for key, value in copy.deepcopy(d).items():
```

Comment:
This is actually editing the passed-in dictionary in-place, which could be a problem if the caller is depending on the dictionary not changing.  Also, I don't think we need to recurse down -- we only need to remove certain keys at the top-level.  Therefore, we could just use a dictionary comprehension (which will make a new dictionary, rather than modifying in-place):

```python
return { k, v for k, v in d.items() if k not in output_params }
```

Reply:
@mdboom We used this util function at various places: 
- if it's to [serialize pings and metrics]( https://github.com/Iinh/glean_parser/blob/0e95f993a4a715c467c405f6b673b04532058b51/glean_parser/metrics.py#L175) then yes we don’t need to recurse down
- however, we also used this function in [`_load_file`](https://github.com/Iinh/glean_parser/blob/0e95f993a4a715c467c405f6b673b04532058b51/glean_parser/parser.py#L93) and in [`load_yaml_or_json`](https://github.com/Iinh/glean_parser/blob/75dfc5e51925adf3bee4bdf763c626b29cb4024c/glean_parser/util.py#L149-L152) to do a bulk removal of all the line numbers because the content passed in is a multi-layered nested dict, so recursion is necessary

I've updated the function to not edit the passed-in dict while still keeping recursion per my explanation above. Looking forward to feedback!